File: scripts/capture-status.py
#!/usr/bin/env python3
# Usage: python3 scripts/capture-status.py <label>
# e.g.   python3 scripts/capture-status.py DK-22251
#        python3 scripts/capture-status.py DK-11201
import sys, time
import usb.core, usb.util

BROTHER_VID = 0x04f9
BROTHER_PID = 0x209d
LABEL = sys.argv[1] if len(sys.argv) > 1 else 'unknown'

BYTE_NAMES = [
    'Print head mark','Size','Fixed B','Device dep','Device dep',
    'Fixed 0x30','Country?','Fixed 0x00',
    'Error info 1','Error info 2',
    'Media width mm','Media type',
    'byte12','byte13','byte14','Mode',
    'byte16','Media length mm','Status type','Phase type',
    'Phase no hi','Phase no lo','Notification','byte23','byte24',
    'byte25','byte26','byte27','byte28','byte29','byte30','byte31',
]

def dump(data, label):
    print(f"\n=== {label} ({len(data)} bytes) ===")
    for i, b in enumerate(data):
        name = BYTE_NAMES[i] if i < len(BYTE_NAMES) else f'byte{i}'
        print(f"  [{i:02d}] 0x{b:02X}  ({b:3d})  {name}")

def drain(ep_in):
    """Read and discard any bytes already sitting in the IN buffer."""
    while True:
        try:
            ep_in.read(64, timeout=100)
        except:
            break

dev = usb.core.find(idVendor=BROTHER_VID, idProduct=BROTHER_PID)
if dev is None:
    print("Device not found — unplug/replug and try again")
    sys.exit(1)

print(f"Found: VID=0x{BROTHER_VID:04x} PID=0x{BROTHER_PID:04x}  roll: {LABEL}")

if dev.is_kernel_driver_active(0):
    dev.detach_kernel_driver(0)

try:
    dev.set_configuration()
except usb.core.USBError as e:
    print(f"set_configuration skipped ({e}) — proceeding anyway")

intf = dev.get_active_configuration()[(0, 0)]
ep_out = usb.util.find_descriptor(intf, custom_match=lambda e: usb.util.endpoint_direction(e.bEndpointAddress) == usb.util.ENDPOINT_OUT)
ep_in  = usb.util.find_descriptor(intf, custom_match=lambda e: usb.util.endpoint_direction(e.bEndpointAddress) == usb.util.ENDPOINT_IN)

# Preamble puts printer in raster mode (needed for status to respond)
ep_out.write(bytes([0x1b, 0x69, 0x61, 0x01]) + bytes(200) + bytes([0x1b, 0x40]) + bytes([0x1b, 0x69, 0x61, 0x01]))
time.sleep(0.5)
drain(ep_in)

# --- 32-byte status ---
ep_out.write(bytes([0x1b, 0x69, 0x53]))
time.sleep(0.8)
try:
    data = ep_in.read(64, timeout=3000)
    dump(data[:32], f"STATUS — {LABEL}")
except Exception as e:
    print(f"STATUS read failed: {e}")

drain(ep_in)

# --- amedia (should be 127 bytes if supported) ---
ep_out.write(bytes([0x1b, 0x69, 0x55, 0x77, 0x01]))
time.sleep(0.8)
try:
    data2 = ep_in.read(256, timeout=3000)
    dump(data2, f"AMEDIA — {LABEL}")
except Exception as e:
    print(f"AMEDIA: no response — {e}")

usb.util.release_interface(dev, 0)
print("\nDone.")

File: scripts/capture-status.mjs
// Captures raw status and amedia bytes from the printer.
// Run: node scripts/capture-status.mjs DK-22251
// Then swap roll and run: node scripts/capture-status.mjs DK-11201
import { createRequire } from 'module';
const require = createRequire(import.meta.url);
const usb = require('/home/mannes/thermal-label/brother-ql/node_modules/.pnpm/usb@2.17.0/node_modules/usb/dist/index.js');

const BROTHER_VID = 0x04f9;
const LABEL = process.argv[2] ?? 'unknown';

const STATUS_REQUEST = Buffer.from([0x1b, 0x69, 0x53]);
const AMEDIA_REQUEST = Buffer.from([0x1b, 0x69, 0x55, 0x77, 0x01]);

const BYTE_NAMES = [
  'Print head mark', 'Size', 'Fixed B', 'Device dep', 'Device dep',
  'Fixed 0x30', 'Country?', 'Fixed 0x00',
  'Error info 1', 'Error info 2',
  'Media width mm', 'Media type',
  'Fixed?', 'Fixed?', 'Reserved?', 'Mode',
  'Fixed?', 'Media length mm', 'Status type', 'Phase type',
  'Phase no hi', 'Phase no lo', 'Notification', 'Reserved', 'Reserved',
];

function hex(buf, label) {
  if (!buf || buf.length === 0) { console.log('  (no data)'); return; }
  console.log(`  ${label} (${buf.length} bytes):`);
  for (let i = 0; i < buf.length; i++) {
    const name = BYTE_NAMES[i] ? ` ← ${BYTE_NAMES[i]}` : '';
    console.log(`  [${String(i).padStart(2, '0')}] 0x${buf[i].toString(16).padStart(2, '0').toUpperCase()}  (${buf[i]})${name}`);
  }
}

const sleep = ms => new Promise(r => setTimeout(r, ms));

function write(ep, data) {
  return new Promise((resolve, reject) =>
    ep.transfer(data, err => (err ? reject(err) : resolve())),
  );
}

function read(ep, len) {
  return new Promise((resolve, reject) =>
    ep.transfer(len, (err, data) => (err ? reject(err) : resolve(data ?? Buffer.alloc(0)))),
  );
}

const dev = usb.getDeviceList().find(d => d.deviceDescriptor.idVendor === BROTHER_VID);
if (!dev) { console.error('No Brother QL printer found'); process.exit(1); }
console.log(`Printer: VID=0x${dev.deviceDescriptor.idVendor.toString(16)} PID=0x${dev.deviceDescriptor.idProduct.toString(16)}  roll: ${LABEL}\n`);

dev.open();
const iface = dev.interface(0);
if (process.platform === 'linux' && iface.isKernelDriverActive()) iface.detachKernelDriver();
iface.claim();
const out = iface.endpoints.find(e => e.direction === 'out');
const inp = iface.endpoints.find(e => e.direction === 'in');

// --- 32-byte status ---
await write(out, STATUS_REQUEST);
await sleep(200);
const statusBytes = await read(inp, 64); // read up to 64 to catch any overflow
hex(statusBytes.slice(0, 32), `STATUS  (${LABEL})`);

await sleep(100);

// --- amedia (127 bytes) ---
await write(out, AMEDIA_REQUEST);
await sleep(300);
let amediaBytes;
try {
  amediaBytes = await read(inp, 256);
  if (amediaBytes.length > 0) {
    hex(amediaBytes, `AMEDIA  (${LABEL})`);
  } else {
    console.log('AMEDIA: no response (command not supported?)');
  }
} catch (e) {
  console.log(`AMEDIA: error — ${e.message}`);
}

iface.release(() => dev.close());
